Add option to treat secrets as mutable

MaienM opened the pull request:
### Proposed changes

This introduces the flag `enableSecretMutable` which does for `Secret`s what the existing `enableConfigMapMutable` flag does for `ConfigMap`s. See #1926 for the motivation for this flag. Changes to the `type` field will still trigger a replacement as this field is immutable.

### Related issues (optional)

Fixes #2291.
Fixes #3181.

blampe commented:
/run-acceptance-tests

blampe commented:
Thank you for the contribution! A couple test failures but looks reasonable overall. We'll review this as a team next week.

```
=== RUN   TestPatchToDiff/Secret_resources_trigger_a_replace_when_enableSecretMutable_is_not_set.
    diff_test.go:331: 
        	Error Trace:	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/diff_test.go:331
        	Error:      	Not equal: 
        	            	expected: map[string]*pulumirpc.PropertyDiff{"data.property1":(*pulumirpc.PropertyDiff)(0xc000893a70)}
        	            	actual  : map[string]*pulumirpc.PropertyDiff{"data.property1":(*pulumirpc.PropertyDiff)(0xc000748360)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -13,3 +13,3 @@
        	            	   unknownFields: ([]uint8) <nil>,
        	            	-  Kind: (pulumirpc.PropertyDiff_Kind) 5,
        	            	+  Kind: (pulumirpc.PropertyDiff_Kind) 4,
        	            	   InputDiff: (bool) false
        	Test:       	TestPatchToDiff/Secret_resources_trigger_a_replace_when_enableSecretMutable_is_not_set.
=== RUN   TestPatchToDiff/Secret_resources_don't_trigger_a_replace_when_mutable.
=== RUN   TestPatchToDiff/Secret_resources_trigger_a_replace_when_type_changes_even_if_enableSecretMutable_is_set.
    diff_test.go:331: 
        	Error Trace:	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/diff_test.go:331
        	Error:      	Not equal: 
        	            	expected: map[string]*pulumirpc.PropertyDiff{"data.property1":(*pulumirpc.PropertyDiff)(0xc000893a70)}
        	            	actual  : map[string]*pulumirpc.PropertyDiff{"type":(*pulumirpc.PropertyDiff)(0xc0007492c0)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,3 @@
        	            	 (map[string]*pulumirpc.PropertyDiff) (len=1) {
        	            	- (string) (len=14) "data.property1": (*pulumirpc.PropertyDiff)({
        	            	+ (string) (len=4) "type": (*pulumirpc.PropertyDiff)({
        	            	   state: (impl.MessageState) {
        	Test:       	TestPatchToDiff/Secret_resources_trigger_a_replace_when_type_changes_even_if_enableSecretMutable_is_set.
```

MaienM commented:
Fixed! One of those was some sloppy copy-pasting in the test itself, the other was an actual logic error that I introduced when I re-marked the `type` field as always immutable.

blampe commented:
/run-acceptance-tests

rquitales reviewed (approved):
LGTM - thanks for the PR!

blampe commented on provider/pkg/provider/diff.go:
```diff
@@ -24,7 +24,9 @@ func (k *kubeProvider) forceNewProperties(obj *unstructured.Unstructured) []stri
 	if group, groupExists := forceNew[gvk.Group]; groupExists {
 		if version, versionExists := group[gvk.Version]; versionExists {
-			if kindFields, kindExists := version[gvk.Kind]; kindExists {
+			if clients.IsSecret(obj) && !k.enableSecretMutable {
+				props = append(props, properties{".type", ".stringData", ".data"}...)
+			} else if kindFields, kindExists := version[gvk.Kind]; kindExists {
```
@MaienM it just occurred to me while adding some detail to https://github.com/pulumi/pulumi-kubernetes/issues/3179 that we should probably respect the `immutable` field.

In other words, if the secret has `immutable: true` it should preserve the existing behavior even if the provider has `enableSecretMutable`. Does that make sense?

MaienM replied:
That makes sense, yeah, it's not like trying to treat it as mutable will even work in that case. The same applies to `ConfigMap`s though, right? I don't think those handle that either right now, that's what I based this code on after all.

I can make a change to handle that, should I also make the same change for `ConfigMap`s while I'm at it?

MaienM replied:
I've made the change for this for both `Secret`s and `ConfigMap`s. I put the change for the `ConfigMap`s in a separate commit because I wasn't sure whether you wanted that one, so you can just get rid of that commit if you don't.

blampe replied:
@MaienM thank you again for the great contribution! I'm very happy this prompted us to uncover the `immutable` incompatibility and that you were able to fix it while you were in here. I added an E2E test case partly as a sanity check and party to cement the behavior we expect here.

So much has changed with Pulumi and k8s since this behavior was originally introduced that it would make much more sense IMO to make this the default behavior at some point in the future (#3179). I added a note to the changelog to that effect, since it's a potentially disruptive behavioral change.

After I merge this it will be available as a pre-release if you'd like to play around with it.

MaienM commented on provider/pkg/clients/clients.go:
```diff
@@ -316,12 +316,20 @@ func IsSecret(obj *unstructured.Unstructured) bool {
 	return (gvk.Group == corev1.GroupName || gvk.Group == "core") && gvk.Kind == string(kinds.Secret)
 }
 
-// IsConfigMap returns true if the resource is a configmap marked as immutable.
```
This is technically unrelated but given that this method doesn't actually consider whether the object is immutable or not it felt misleading, so I figured I'd get rid of it.

blampe commented:
/run-acceptance-tests
